src/Control/Monad/Catch.hs

```diff
@@ -194,61 +214,95 @@ class MonadCatch m => MonadMask m where
+  -- returned by the @use@ and @release@ actions. The effect part of the @use@
+  -- result, in this case @_s2@, usually needs to be discarded, since those
+  -- effects have already been incorporated in the @release@ action.
+  --
+  -- The only effect which is intentionally not incorporated in the @release@
+  -- action is the effect of throwing an error. In that case, the error must be
+  -- re-thrown. One subtelty which is easy to miss is that in the case in which
+  -- @use@ and @release@ both throw an error, the error from @release@ should
+  -- take priority. Here is an implementation for @ExceptT@ which demonstrates
+  -- how to do this.
   --
   -- @
-  -- generalBracket acquire release cleanup use = ExceptT $
-  --   generalBracket
+  -- generalBracket acquire use release = ExceptT $ do
+  --   (eb, ec) <- generalBracket
   --     (runExceptT acquire)
-  --     (\eresource ->
-  --       case eresource of
-  --         Left _ -> return ()
-  --         Right resource -> runExceptT (release resource) >> return ())
-  --     (\eresource e ->
-  --        case eresource of
-  --          Left _ -> return ()
-  --          Right resource -> runExceptT (cleanup resource e) >> return ())
   --     (either (return . Left) (runExceptT . use))
+  --     (\eresource exitCase -> case eresource of
+  --       Left e -> return (Left e) -- nothing to release, acquire didn't succeed
+  --       Right resource -> case exitCase of
+  --         ExitCaseSuccess (Right b) -> runExceptT (release resource (ExitCaseSuccess b))
+  --         ExitCaseException e       -> runExceptT (release resource (ExitCaseException e))
+  --         _                         -> runExceptT (release resource ExitCaseAbort))
+  --   return $ do
+  --     -- The order in which we perform those two 'Either' effects determines
+  --     -- which error will win if they are both 'Left's. We want the error from
+  --     -- 'release' to win.
+  --     c <- ec
+  --     b <- eb
+  --     return (b, c)
   -- @
   --
-  -- In this implementation, we need to deal with the potential that
-  -- the @acquire@ action returned a @Left@ (as opposed to succeeding
-  -- with a @Right@ or throwing an exception via @throwM@), and
-  -- therefore have to handle the @Left@ case explicitly when provide
-  -- @release@, @cleanup@, and @use@ actions to the base monad's
-  -- implementation of @generalBracket@.
-  --
-  -- You should ensure that in all cases of the @acquire@ action
-  -- completing successfully, either the @release@ or @cleanup@
-  -- actions are called, regardless of what occurs in @use@.
-  --
   -- @since 0.9.0
   generalBracket
     :: m a
     -- ^ acquire some resource
-    -> (a -> m ignored1)
-    -- ^ release, no exception thrown
-    -> (a -> SomeException -> m ignored2)
-    -- ^ release, some exception thrown; the exception will be rethrown
     -> (a -> m b)
     -- ^ inner action to perform with the resource
-    -> m b
+    -> (a -> ExitCase b -> m c)
```

Comment:
I should have commented on this earlier. I think it would make more sense to move the release action to before the use action, to match more closely the type signature of `bracket` in `base`.